File: csharp/tests/Idkollen.Client.Tests/VippsTests.cs
using System.Net;
using System.Text.Json;
using Idkollen.Client.Models;
using Xunit;

namespace Idkollen.Client.Tests;

public class VippsTests
{
    private static (Idkollen.Client.IdkollenClient Client, MockHttpMessageHandler Handler) Build()
    {
        var handler = new MockHttpMessageHandler();
        var http = new HttpClient(handler);
        var client = new Idkollen.Client.IdkollenClientBuilder("cid", "sec")
            .BaseUrl("https://x.test").HttpClient(http).Build();
        return (client, handler);
    }

    [Fact]
    public async Task AuthPending()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { status = "PENDING", id = "v1", url = "https://login" });
        var result = await client.Vipps.AuthAsync(new VippsAuthRequest());
        Assert.IsType<VippsPending>(result);
    }

    [Fact]
    public async Task BackchannelAuthSendsPhone()
    {
        var (client, handler) = Build();
        handler.Enqueue(HttpStatusCode.OK, new { status = "PENDING", id = "v1" });
        await client.Vipps.BackchannelAuthAsync(new VippsBackchannelAuthRequest("+4712345678"));
        using var doc = JsonDocument.Parse(handler.LastRequestBody);
        Assert.Equal("+4712345678", doc.RootElement.GetProperty("phone").GetString());
    }
}

File: swift/Tests/IdkollenClientTests/VippsTests.swift
import XCTest
@testable import IdkollenClient

final class VippsTests: XCTestCase {
    var client: IdkollenClient!

    override func setUp() {
        super.setUp()
        MockURLProtocol.reset()
        client = IdkollenClientBuilder(clientId: "cid", clientSecret: "sec")
            .baseUrl("https://x.test")
            .urlSession(.mocked())
            .build()
    }

    func testAuthPending() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "PENDING",
            "id": "abc",
            "url": "https://login",
        ])

        let result = try await client.vipps.auth(VippsAuthRequest())
        guard case .pending(let p) = result else { return XCTFail("expected pending") }
        XCTAssertEqual(p.url, "https://login")
        XCTAssertEqual(p.id, "abc")
    }

    func testBackchannelAuthSendsPhone() async throws {
        MockURLProtocol.enqueue(status: 200, json: [
            "status": "PENDING",
            "id": "abc",
        ])

        _ = try await client.vipps.backchannelAuth(VippsBackchannelAuthRequest(phone: "+4712345678"))

        let body = try JSONSerialization.jsonObject(with: MockURLProtocol.capturedBodies.last!) as! [String: Any]
        XCTAssertEqual(body["phone"] as? String, "+4712345678")
    }
}
